core/block.ts

```diff
@@ -2084,16 +2086,64 @@ export class Block implements IASTNodeLocation, IDeletable {
    * @param _opt_id An optional ID for the warning text to be able to maintain
    *     multiple warnings.
    */
-  setWarningText(_text: string|null, _opt_id?: string) {}
-  // NOP.
+  setWarningText(_text: string|null, _opt_id?: string) {
+    // NOOP.
+  }
 
   /**
    * Give this block a mutator dialog.
    *
    * @param _mutator A mutator dialog instance or null to remove.
    */
-  setMutator(_mutator: Mutator) {}
-  // NOP.
+  setMutator(_mutator: Mutator) {
+    // NOOP.
+  }
+
+  /** Adds the given icon to the block. */
+  addIcon<T extends IIcon>(icon: T): T {
+    if (this.hasIcon(icon.getType())) {
+      throw new Error(
+          'Tried to append an icon when an icon of the same type already ' +
+          'exists on the block. Use getIcon to access the existing icon.');
+    }
+    this.icons.push(icon);
+    this.icons.sort((a, b) => a.getWeight() - b.getWeight());
+    return icon;
+  }
+
+  /**
+   * Removes the icon whose getType matches the given type iconType from the
+   * block.
+   *
+   * @param type The type of the icon to remove from the block.
+   * @return True if an icon with the given type was found, false otherwise.
+   */
+  removeIcon(type: string): boolean {
+    if (!this.hasIcon(type)) return false;
```

Comment:
I sort of think removing an icon that doesn't exist should throw an error the same way adding one that already does throws. It feels more consistent. But I think you did it this way to be consistent with fields?

Reply:
[`removeField`](https://github.com/google/blockly/blob/9442b73d99fb63da4526c651ac41762e458b4ed7/core/inputs/input.ts#L145) optionally throws an error (which I think is a weird and bad pattern). I had it return a boolean to be more consistent with other ecmascript things (e.g. maps).